vpnglean/glean_parser_ext/templates/cpp.jinja2

```diff
@@ -38,28 +38,34 @@ struct {{ obj.name|Camelize }}{{ suffix }} {
   {% endfor %}
 
   FfiExtra ToFfiExtra() const {
-    QVector<const char*> extraKeys;
-    QVector<const char*> extraValues;
+    const char* extraKeys[{{ obj|attr(name)|length }} + 1];
+    const char* extraValues[{{ obj|attr(name)|length }} + 1];
+    int count = 0;
+
+    QVector<QByteArray> _keepStringsAlive;
     {% for item, type in obj|attr(name) %}
     if (_{{item|camelize}}.canConvert<{{type|extra_type_name}}>()) {
-      extraKeys.append("{{item}}");
+      extraKeys[count] = "{{item}}";
       {% if type == "string" %}
-      extraValues.append(_{{item|camelize}}.toString().toUtf8().constData());
+      _keepStringsAlive.append(_{{item|camelize}}.toString().toUtf8());
+      extraValues[count] = _keepStringsAlive[count].constData();
       {% elif type == "boolean" %}
-      extraValues.append(_{{item|camelize}}.toBool() ? "true" : "false");
+      _keepStringsAlive.append(_{{item|camelize}}.toBool() ? "true" : "false");
+      extraValues[count] = _keepStringsAlive[count];
       {% elif type == "quantity" %}
-      std::string s = std::to_string(_{{item|camelize}}.toInt());
-      extraValues.append(s.c_str());
+      _keepStringsAlive.append(QString::number(_{{item|camelize}}.toInt()).toUtf8());
+      extraValues[count] = _keepStringsAlive[count].constData();
       {% else %}
 #error "Glean: Invalid extra key type for metric {{obj.category}}.{{obj.name}}, defined in: {{obj.defined_in['filepath']}}:{{obj.defined_in['line']}})"
       {% endif %}
+      count++;
     }
     {% endfor %}
 
     FfiExtra extras;
-    extras.keys = std::move(extraKeys.begin());
-    extras.values = std::move(extraValues.begin());
-    extras.count = extraValues.count();
+    extras.keys = std::move(extraKeys);
+    extras.values = std::move(extraValues);
+    extras.count = count;
```

Comment:
This will not work because at this point, `keepStringAlive` will be out-of-scope.

Reply:
But this is happening at the same scope as keepStringsAlive :thinking:

Reply:
Here is an example of the generated code:
```cpp
FfiExtra ToFfiExtra() const {
  const char* extraKeys[2 + 1];
  const char* extraValues[2 + 1];
  int count = 0;

  QVector<QByteArray> _keepStringsAlive;
  if (_addonId.canConvert<QString>()) {
    extraKeys[count] = "addon_id";
    _keepStringsAlive.append(_addonId.toString().toUtf8());
    extraValues[count] = _keepStringsAlive[count].constData();
    count++;
  }
  if (_state.canConvert<QString>()) {
    extraKeys[count] = "state";
    _keepStringsAlive.append(_state.toString().toUtf8());
    extraValues[count] = _keepStringsAlive[count].constData();
    count++;
  }

  FfiExtra extras;
  extras.keys = std::move(extraKeys);
  extras.values = std::move(extraValues);
  extras.count = count;
  return extras;
};
```

Reply:
Right. The result is an array of char* that are not kept alive outside this method. At the end of the method, keepAlive is released and the pointers are freed.

Reply:
Ugh, duh. Ok So I need to put that in the outer scope. Thanks for the patience in walking me through this one @bakulf :D

Reply:
Tried again, this is what the generated code looks like now:
```cpp
   struct AddonMessageStateChangedExtra {
      QVariant _messageId = QVariant();
      QVariant _messageState = QVariant();

      FfiExtra ToFfiExtra() const {
        m_keepStringsAlive.clear();

        const char* extraKeys[2 + 1];
        const char* extraValues[2 + 1];
        int count = 0;

        if (_messageId.canConvert<QString>()) {
          QByteArray s = _messageId.toString().toUtf8();
          m_keepStringsAlive.append(s);
          extraValues[count] = s.constData();
          extraKeys[count] = "message_id";
          count++;
        }
        if (_messageState.canConvert<QString>()) {
          QByteArray s = _messageState.toString().toUtf8();
          m_keepStringsAlive.append(s);
          extraValues[count] = s.constData();
          extraKeys[count] = "message_state";
          count++;
        }

        FfiExtra extras;
        extras.keys = std::move(extraKeys);
        extras.values = std::move(extraValues);
        extras.count = count;
        return extras;
      };

      private:
        // Helper vector to extend the lifetime of the strings
        // which hold the extra key values until they are used.
        QVector<QByteArray> m_keepStringsAlive;
    };
```